src/back/glsl.rs

```diff
@@ -75,18 +76,193 @@ pub struct TextureMapping {
     pub sampler: Option<Handle<GlobalVariable>>,
 }
 
-const SUPPORTED_CORE_VERSIONS: &[u16] = &[450, 460];
+const SUPPORTED_CORE_VERSIONS: &[u16] = &[330, 400, 410, 420, 430, 440, 450];
 const SUPPORTED_ES_VERSIONS: &[u16] = &[300, 310];
 
 bitflags::bitflags! {
-    struct SupportedFeatures: u32 {
+    struct Features: u32 {
         const BUFFER_STORAGE = 1;
-        const SHARED_STORAGE = 1 << 1;
+        const ARRAY_OF_ARRAYS = 1 << 1;
         const DOUBLE_TYPE = 1 << 2;
-        const NON_FLOAT_MATRICES = 1 << 3;
+        const FULL_IMAGE_FORMATS = 1 << 3;
         const MULTISAMPLED_TEXTURES = 1 << 4;
         const MULTISAMPLED_TEXTURE_ARRAYS = 1 << 5;
-        const NON_2D_TEXTURE_ARRAYS = 1 << 6;
+        const CUBE_TEXTURES_ARRAY = 1 << 6;
+        const COMPUTE_SHADER = 1 << 7;
+        const IMAGE_LOAD_STORE = 1 << 8;
+        const CONSERVATIVE_DEPTH = 1 << 9;
+        const TEXTURE_1D = 1 << 10;
+    }
+}
+
+struct FeaturesManager(Features);
+
+impl FeaturesManager {
+    pub fn new() -> Self {
+        Self(Features::empty())
+    }
+
+    pub fn request(&mut self, features: Features) {
+        self.0 |= features
+    }
+
+    #[allow(clippy::collapsible_if)]
+    pub fn write(&self, version: Version, out: &mut impl Write) -> Result<(), Error> {
+        let (v, es) = match version {
+            Version::Desktop(v) => (v, false),
+            Version::Embedded(v) => (v, true),
+        };
+
+        if self.0.contains(Features::COMPUTE_SHADER) {
+            if (es && v < 310) || (!es && v < 420) {
```

Comment:
This may be a bit error-prone. It would better to encode these restrictions as `Version` values, basically, and then define a `PartialOrd` for them in a way that `Desktop` vs `Embedded` is considered to have no order. See https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=276f348602ba91fb57265ce50d524bc4 for inspiration